[WIP] Separator: Trial custom separator height with independent top/bottom margins

aaronrobertshaw opened the pull request:
## Description
Trial exploring splitting a custom separator height into both top and bottom margins, while still providing the ability to drag a resizable box handle to visually set the height. 

This PR is based off https://github.com/WordPress/gutenberg/pull/28409

* This does not extend the spacing block support to provide margins control
* It does use `style.spacing.margin.top` and `style.spacing.margin.bottom` attributes to store the values
* Should the spacing block support get the margins feature merged this should be a simple opt in, then remove the current margin controls.

#### Next Steps
- [ ] Fix bug in native controls where after default values are applied after unit change, the input doesn't reflect the new value.

## How has this been tested?
Manually.

#### Testing Instructions

TBA

## Screenshots <!-- if applicable -->
<img src="https://user-images.githubusercontent.com/60436221/106734517-d6b3d200-665e-11eb-80db-fcaf4edac19f.gif" width="400" />


## Types of changes
New feature.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

aaronrobertshaw committed:
Trial splitting separator margins top/bottom

aaronrobertshaw committed:
Update native separator controls for split margins

stokesman commented:
I like that this approach should ease and make cleaner (no attribute deprecation) a transition to stock spacing controls for margins if they do ever land and as you said in the other PR:
>assuming the decision to opt-in to spacing margins support is made.

Since there are two levels of uncertainty there and because I was curious to try it, I messed around with alternative inspector controls (basing off this PR). I made something that goes like this:

https://user-images.githubusercontent.com/9000376/107415598-6971c800-6ac8-11eb-9532-b0d8a34e18d7.mp4

I think some of it may be good (probably not the arrow icons). It does mean sharing the unit across top and bottom, unlike the PR at present. The odd thing implementation-wise is it's not clear that such customization of inspector controls is precedented in blocks (I'm not sure that the CSS for such is acceptable in the block editor styles) so maybe the `UnitControl` and `RangeControl` should be kept to their own rows.

A newer idea I had was maybe **two** resize handles could be employed. Though that'd probably want the presence of the second to be controlled via a lock/unlock toggle for the side controls in the inspector.

All this could be taking the exercise further than needed for a feature that I still imagine to be niche. If the use cases you have in mind @aaronrobertshaw, can all be met with the simple evenly distributed margin I'd say your former PR is fine. Though maybe it'd be finer also using the style/spacing attributes 🤷‍♂️ .

aaronrobertshaw committed:
Assign min height to allow immediate drag resizing

aaronrobertshaw commented:
Thanks for the extra feedback 👍 

There's definitely an opportunity to refine the controls once a decision is made on general approach. We need to consider the controls for native as well.

> It does mean sharing the unit across top and bottom, unlike the PR at present. 

I'm not sure I follow, this PR does store different top and bottom margin values. If there is a single control field, that value can be split into the two attributes, similar to the way that the singular height value from the resizable box gets split now.

> A newer idea I had was maybe **two** resize handles could be employed. Though that'd probably want the presence of the second to be controlled via a lock/unlock toggle for the side controls in the inspector.

It could be possible to do something here like that. I had avoided it previously in an attempt to keep more consistency with the spacer block.

Another possible concern is adding another layer of complexity to the usability and intuitiveness of the controls. As you suggested, perhaps if the second drag handle was locked behind a control toggle or something it might help.

> All this could be taking the exercise further than needed for a feature that I still imagine to be niche. 

I believe there will be plenty of use cases for designers to utilise custom separator heights within block patterns to be used across themes.  That said, I agree adding a second drag handle for now might be a step too far. It could easily be revisited in the future.

> height does not update while dragging (unlike in the Spacer block).

I've fixed the issue preventing drag resizing on first insertion of the block as mentioned in https://github.com/WordPress/gutenberg/pull/28409#issuecomment-776227037. At this stage, I had opted out of updating the height `onResize` as it was performing rather poorly. It is only my list of things to address when a decision between the two approaches is made and I refactor/polish up the winner ✨

stokesman commented:
>I'm not sure I follow, this PR does store different top and bottom margin values.

Right, I was contrasting this PR with the version seen in the video that has only one `UnitControl` so the value is split, but the unit is shared.

aaronrobertshaw committed:
Update margins on box resizing

When dev tools is open this is pretty jumpy but others feels smooth enough.

aaronrobertshaw committed:
Unify min margins to simplify PR and refactor

- Adjusts margin minimums so they are the same regardless of block style
  such as Dots.
- Refactored edit component moving helpers into shared file to clean up

aaronrobertshaw commented:
This has been updated so that height is adjusted on resize rather than when the resizing stops.

In addition, I've made the minimum margin defaults the same regardless of block style. Following that this PR has had a bit of a refactor to clean things up and make it a little easier to follow.

One issue still outstanding is with the `UnitControl` on native. Currently, when the unit selection is changed, the margin values are reset to the default for that unit. Unfortunately, I'm missing something and the `UnitControl`'s field value isn't updating to match the default value set.

<img src="https://user-images.githubusercontent.com/60436221/107609239-18113680-6c8a-11eb-870b-15ac8a59b32d.gif" width="300" />

aaronrobertshaw committed:
Fix issue switching units and setting defaults via native unit controls

aaronrobertshaw committed:
Cleanup onUnitChange and remove obsolete comment

apeatling commented:
I think we should concentrate on https://github.com/WordPress/gutenberg/pull/29363 and look to add the handle in a future iteration.

aaronrobertshaw commented:
Closing this in favour of getting #29363 merged as first iteration.
